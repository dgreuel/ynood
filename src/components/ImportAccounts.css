.import-form {
  text-align: center;
}

.import-form input {
  width: 100px;
}

.import-form label {
  padding: 3px;
}

.import-form-account {
  display: flex;
  flex-direction: row;
  border: 1px dashed black;
  margin-bottom: 5px;
}

.import-form-account .account-left {
  padding: 3px;
  display: flex;
  flex-direction: column;
  margin: 10px;
}

.import-form-account .account-right {
  text-align: left;
}

.import-form .balance {
  font-style: italic;
}
