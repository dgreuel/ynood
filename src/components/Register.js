import { Form, Text } from 'informed'
import React from 'react'
import { Container, Col, Row } from 'mdbreact'

export default class Register extends React.Component {
  render() {
    return (
      <div className="register">
        <p className="h5 text-center mb-4">
          Step 2: Register for a YNOOD Account
        </p>
        <Container>
          <Row>
            <Col md="1" />
            <Col md="10">
              <Form
                id="register-form"
                onSubmit={values =>
                  this.props.registrationFunction(
                    values.email,
                    this.props.ynabID,
                    values.budget
                  )
                }
              >
                {({ formApi }) => (
                  <div>
                    <label htmlFor="register-email">Email:</label>
                    <Text
                      field="email"
                      id="register-email"
                      className={
                        formApi.getState().errors.email
                          ? 'is-invalid form-control'
                          : 'form-control'
                      }
                      validate={validateEmail}
                      validateOnBlur
                    />
                    <br />
                    <label htmlFor="register-budget">
                      Monthly debt payoff budget:
                    </label>
                    <Text
                      field="budget"
                      id="register-budget"
                      className={
                        formApi.getState().errors.budget
                          ? 'is-invalid form-control'
                          : 'form-control'
                      }
                      validate={validateBudget}
                      validateOnBlur
                      validateOnChange
                    />
                  </div>
                )}
              </Form>
            </Col>
            <Col md="1" />
          </Row>
          <Row>
            <Col md="1" />
            <Col md="10">
              <button type="submit" className="btn btn-outline-purple">
                Submit
              </button>
            </Col>
            <Col md="1" />
          </Row>
        </Container>
      </div>
    )
  }
}
const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
const budgetRegex = /^[1-9][0-9]*(\.[0-9]{2})?$/

const validateEmail = value => {
  return typeof value === 'string'
    ? emailRegex.test(value)
      ? null
      : 'Please enter a valid email address'
    : 'Invalid email input.'
}

const validateBudget = value => {
  return typeof value === 'string'
    ? budgetRegex.test(value)
      ? null
      : 'Please enter a valid dollar amount'
    : 'Invalid budget input.'
}
