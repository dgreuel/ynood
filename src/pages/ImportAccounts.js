import { Form, Text, Scope } from 'informed'
import React from 'react'
import { Container, Col, Row } from 'mdbreact'

export default class ImportAccounts extends React.Component {
  render() {
    return (
      <div className="import">
        <p className="h5 text-center mb-4">Step 3: Enter Debt Details</p>
        <Container>
          <Row>
            <Col md="12">
              <Form
                id="import-form"
                onSubmit={values => console.log(values)// this.props.importFunction(values)
                }
              >
                {({ formApi }) => (
                  <div>
                    {this.props.accounts.map((account, index, array) => {
                      return (
                        <Scope scope={'accounts[' + index + ']'}>
                          <label htmlFor={'account-' + index + '-interestrate'}>
                            Interest Rate
                          </label>
                          <Text
                            field="interestrate"
                            id={'account-' + index + '-interestrate'}
                            key={'account-' + index + '-interestrate'}
                            className={
                              errorExists(
                                formApi.getState().errors,
                                'accounts',
                                index,
                                'interestrate'
                              )
                                ? 'is-invalid form-control'
                                : 'form-control'
                            }
                            validate={validateInterestRate}
                            validateOnBlur
                          />
                          <label htmlFor={'account-' + index + '-nextduedate'}>
                            Next Due Date
                          </label>
                          <Text
                            field="nextduedate"
                            id={'account-' + index + '-nextduedate'}
                            key={'account-' + index + '-nextduedate'}
                            className={
                              formApi.getState().errors.nextduedate
                                ? 'is-invalid form-control'
                                : 'form-control'
                            }
                            validate={validateDueDate}
                            validateOnBlur
                          />
                        </Scope>
                      )
                    })}
                    <button type="submit" className="btn btn-outline-purple">
                      Submit
                    </button>
                  </div>
                )}
              </Form>
            </Col>
          </Row>
        </Container>
      </div>
    )
  }
}

const errorExists = (errors, arrayName, index, field) => {
  if (errors && errors[arrayName] && errors[arrayName].length) {
    // console.log('returning ' + arrayName + '-' + index + '-' + field)
    return errors[arrayName][index] ? errors[arrayName][index][field] : false
  }
  return false
}
const interestRateRegex = /[0-9]+(\.[0-9]+)?/

const validateInterestRate = value => {
  return typeof value === 'string'
    ? interestRateRegex.test(value)
      ? null
      : 'Please enter a valid interest rate in %'
    : 'Invalid interest rate input.'
}

const validateDueDate = value => {
  return null // check if date is in the future using moment?
}
