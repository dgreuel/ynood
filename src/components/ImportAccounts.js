import { Form, Text, Scope } from 'informed'
import React from 'react'
import { Container, Col, Row } from 'mdbreact'
import { isDebtAccount } from '../pages/Home'
import * as accounting from 'accounting'
import './ImportAccounts.css'

export default class ImportAccounts extends React.Component {
  render() {
    const { userID, importFunction, fetchFunction } = this.props
    return (
      <div className="import">
        <p className="h5 text-center mb-4">Step 3: Enter Debt Details</p>
        <Container>
          <Row>
            <Col md="12">
              <Form
                id="import-form"
                onSubmit={values =>
                  importFunction(values, userID).then(result =>
                    fetchFunction(userID)
                  )
                }
              >
                {({ formApi }) => (
                  <div className="import-form">
                    {this.props.accounts
                      .filter(account => {
                        return isDebtAccount(account)
                      })
                      .map((account, index, array) => {
                        return (
                          <div
                            className="import-form-account"
                            key={'accounts[' + index + ']'}
                          >
                            <Scope scope={'accounts[' + index + ']'}>
                              <div className="account-left">
                                <Text
                                  field="id"
                                  id={'account-' + index + '-id'}
                                  key={'account-' + index + '-id'}
                                  initialValue={account.id}
                                  type="hidden"
                                />
                                <Text
                                  field="balance"
                                  id={'account-' + index + '-balance'}
                                  key={'account-' + index + '-balance'}
                                  initialValue={account.balance / -1000}
                                  type="hidden"
                                />
                                <Text
                                  field="name"
                                  id={'account-' + index + '-name'}
                                  key={'account-' + index + '-name'}
                                  initialValue={account.name}
                                  type="hidden"
                                />
                                <div>{account.name}</div>
                                <div className="balance">
                                  {accounting.formatMoney(
                                    account.balance / -1000
                                  )}
                                </div>
                              </div>
                              <div className="account-right">
                                <div>
                                  <label
                                    className="import-form-label"
                                    htmlFor={
                                      'account-' + index + '-interestrate'
                                    }
                                  >
                                    Interest Rate
                                  </label>
                                  <Text
                                    field="interestrate"
                                    id={'account-' + index + '-interestrate'}
                                    key={'account-' + index + '-interestrate'}
                                    className={
                                      errorExists(
                                        formApi.getState().errors,
                                        'accounts',
                                        index,
                                        'interestrate'
                                      )
                                        ? 'is-invalid'
                                        : null
                                    }
                                    validate={validateInterestRate}
                                    validateOnBlur
                                  />
                                </div>
                                <div>
                                  <label
                                    htmlFor={
                                      'account-' + index + '-interestrate'
                                    }
                                  >
                                    Min. Monthly Payment
                                  </label>
                                  <Text
                                    field="minimumpayment"
                                    id={'account-' + index + '-minimumpayment'}
                                    key={'account-' + index + '-minimumpayment'}
                                    className={
                                      errorExists(
                                        formApi.getState().errors,
                                        'accounts',
                                        index,
                                        'minimumpayment'
                                      )
                                        ? 'is-invalid'
                                        : null
                                    }
                                    validate={validateMinimumPayment}
                                    validateOnBlur
                                  />
                                </div>
                                <div>
                                  <label
                                    htmlFor={
                                      'account-' + index + '-nextduedate'
                                    }
                                  >
                                    Next Due Date
                                  </label>
                                  <Text
                                    field="nextduedate"
                                    id={'account-' + index + '-nextduedate'}
                                    key={'account-' + index + '-nextduedate'}
                                    className={
                                      formApi.getState().errors.nextduedate
                                        ? 'is-invalid'
                                        : null
                                    }
                                    validate={validateDueDate}
                                    validateOnBlur
                                  />
                                </div>
                              </div>
                            </Scope>
                          </div>
                        )
                      })}
                    <button type="submit" className="btn btn-outline-purple">
                      Submit
                    </button>
                  </div>
                )}
              </Form>
            </Col>
          </Row>
        </Container>
      </div>
    )
  }
}

const errorExists = (errors, arrayName, index, field) => {
  if (errors && errors[arrayName] && errors[arrayName].length) {
    // console.log('returning ' + arrayName + '-' + index + '-' + field)
    return errors[arrayName][index] ? errors[arrayName][index][field] : false
  }
  return false
}
const interestRateRegex = /[0-9]+(\.[0-9]+)?/
const minimumPaymentRegex = /^[1-9][0-9]*(\.[0-9]{2})?$/

const validateInterestRate = value => {
  return typeof value === 'string'
    ? interestRateRegex.test(value)
      ? null
      : 'Please enter a valid interest rate in %'
    : 'Invalid interest rate input.'
}
const validateMinimumPayment = value => {
  return typeof value === 'string'
    ? minimumPaymentRegex.test(value)
      ? null
      : 'Please enter a valid minimum payment in dollars'
    : 'Invalid minimum payment input.'
}
const validateDueDate = value => {
  return null // check if date is in the future using moment?
}
