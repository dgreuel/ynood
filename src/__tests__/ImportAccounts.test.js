import { ImportAccounts } from '../components/ImportAccounts'
import React from 'react'
import 'jest-dom/extend-expect'
import { mockYnabBudget, mockYnoodUser } from './MockData'
import renderer from 'react-test-renderer'

describe('ImportAccounts component', () => {
  const mockImportFunction = jest.fn().mockName('mockImportFunction')
  const mockFetchFunction = jest.fn().mockName('mockFetchFunction')
  const props = {
    accounts: mockYnabBudget.data.budget.accounts,
    importFunction: mockImportFunction,
    fetchFunction: mockFetchFunction,
    userID: mockYnoodUser.undebt_user_id
  }
  beforeAll(() => {})

  beforeEach(() => {})

  it('renders correctly', () => {
    const tree = renderer.create(<ImportAccounts {...props} />)
    expect(tree).toMatchSnapshot()
  })
})
