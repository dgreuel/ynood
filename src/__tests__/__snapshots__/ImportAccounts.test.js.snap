// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportAccounts component renders correctly 1`] = `
<div
  className="import"
>
  <p
    className="h5 text-center mb-4"
  >
    Step 3: Enter Debt Details
  </p>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <form
          id="import-form"
          onReset={[Function]}
          onSubmit={[Function]}
        >
          <div
            className="import-form"
          >
            <div
              className="import-form-account"
            >
              <div
                className="account-left"
              >
                <input
                  id="account-0-id"
                  name="id"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value="575ab9b3-d2cc-4e9b-a7fa-c1b9caf937b0"
                />
                <input
                  id="account-0-balance"
                  name="balance"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value={2342.33}
                />
                <input
                  id="account-0-name"
                  name="name"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value="American Express"
                />
                <div>
                  American Express
                </div>
                <div
                  className="balance"
                >
                  $2,342.33
                </div>
              </div>
              <div
                className="account-right"
              >
                <div>
                  <label
                    className="import-form-label"
                    htmlFor="account-0-interestrate"
                  >
                    Interest Rate
                  </label>
                  <input
                    className={null}
                    id="account-0-interestrate"
                    name="interestrate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
                <div>
                  <label
                    htmlFor="account-0-interestrate"
                  >
                    Min. Monthly Payment
                  </label>
                  <input
                    className={null}
                    id="account-0-minimumpayment"
                    name="minimumpayment"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
                <div>
                  <label
                    htmlFor="account-0-nextduedate"
                  >
                    Next Due Date
                  </label>
                  <input
                    className={null}
                    id="account-0-nextduedate"
                    name="nextduedate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              className="import-form-account"
            >
              <div
                className="account-left"
              >
                <input
                  id="account-1-id"
                  name="id"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value="4f8978af-dc34-4c45-bfda-5fb82925cb93"
                />
                <input
                  id="account-1-balance"
                  name="balance"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value={135789}
                />
                <input
                  id="account-1-name"
                  name="name"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value="Mortgage"
                />
                <div>
                  Mortgage
                </div>
                <div
                  className="balance"
                >
                  $135,789.00
                </div>
              </div>
              <div
                className="account-right"
              >
                <div>
                  <label
                    className="import-form-label"
                    htmlFor="account-1-interestrate"
                  >
                    Interest Rate
                  </label>
                  <input
                    className={null}
                    id="account-1-interestrate"
                    name="interestrate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
                <div>
                  <label
                    htmlFor="account-1-interestrate"
                  >
                    Min. Monthly Payment
                  </label>
                  <input
                    className={null}
                    id="account-1-minimumpayment"
                    name="minimumpayment"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
                <div>
                  <label
                    htmlFor="account-1-nextduedate"
                  >
                    Next Due Date
                  </label>
                  <input
                    className={null}
                    id="account-1-nextduedate"
                    name="nextduedate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              className="import-form-account"
            >
              <div
                className="account-left"
              >
                <input
                  id="account-2-id"
                  name="id"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value="4f8978af-dc34-4c45-bfda-5fb82925cb93-unsynced"
                />
                <input
                  id="account-2-balance"
                  name="balance"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value={125789}
                />
                <input
                  id="account-2-name"
                  name="name"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="hidden"
                  value="Mortgage"
                />
                <div>
                  Mortgage
                </div>
                <div
                  className="balance"
                >
                  $125,789.00
                </div>
              </div>
              <div
                className="account-right"
              >
                <div>
                  <label
                    className="import-form-label"
                    htmlFor="account-2-interestrate"
                  >
                    Interest Rate
                  </label>
                  <input
                    className={null}
                    id="account-2-interestrate"
                    name="interestrate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
                <div>
                  <label
                    htmlFor="account-2-interestrate"
                  >
                    Min. Monthly Payment
                  </label>
                  <input
                    className={null}
                    id="account-2-minimumpayment"
                    name="minimumpayment"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
                <div>
                  <label
                    htmlFor="account-2-nextduedate"
                  >
                    Next Due Date
                  </label>
                  <input
                    className={null}
                    id="account-2-nextduedate"
                    name="nextduedate"
                    onBlur={[Function]}
                    onChange={[Function]}
                    value=""
                  />
                </div>
              </div>
            </div>
            <button
              className="btn btn-outline-purple"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
------------HTML PREVIEW---------------
<div class="import">
  <p class="h5 text-center mb-4">
    Step 3: Enter Debt Details
  </p>
  <div class="container">
    <div class="row">
      <div class="col-md-12">
        <form id="import-form">
          <div class="import-form">
            <div class="import-form-account">
              <div class="account-left">
                <input
                  id="account-0-id"
                  name="id"
                  type="hidden"
                  value="575ab9b3-d2cc-4e9b-a7fa-c1b9caf937b0"
                >
                <input
                  id="account-0-balance"
                  name="balance"
                  type="hidden"
                >
                <input
                  id="account-0-name"
                  name="name"
                  type="hidden"
                  value="American Express"
                >
                <div>
                  American Express
                </div>
                <div class="balance">
                  $2,342.33
                </div>
              </div>
              <div class="account-right">
                <div>
                  <label
                    class="import-form-label"
                    for="account-0-interestrate"
                  >
                    Interest Rate
                  </label>
                  <input
                    id="account-0-interestrate"
                    name="interestrate"
                    value=""
                  >
                </div>
                <div>
                  <label for="account-0-interestrate">
                    Min. Monthly Payment
                  </label>
                  <input
                    id="account-0-minimumpayment"
                    name="minimumpayment"
                    value=""
                  >
                </div>
                <div>
                  <label for="account-0-nextduedate">
                    Next Due Date
                  </label>
                  <input
                    id="account-0-nextduedate"
                    name="nextduedate"
                    value=""
                  >
                </div>
              </div>
            </div>
            <div class="import-form-account">
              <div class="account-left">
                <input
                  id="account-1-id"
                  name="id"
                  type="hidden"
                  value="4f8978af-dc34-4c45-bfda-5fb82925cb93"
                >
                <input
                  id="account-1-balance"
                  name="balance"
                  type="hidden"
                >
                <input
                  id="account-1-name"
                  name="name"
                  type="hidden"
                  value="Mortgage"
                >
                <div>
                  Mortgage
                </div>
                <div class="balance">
                  $135,789.00
                </div>
              </div>
              <div class="account-right">
                <div>
                  <label
                    class="import-form-label"
                    for="account-1-interestrate"
                  >
                    Interest Rate
                  </label>
                  <input
                    id="account-1-interestrate"
                    name="interestrate"
                    value=""
                  >
                </div>
                <div>
                  <label for="account-1-interestrate">
                    Min. Monthly Payment
                  </label>
                  <input
                    id="account-1-minimumpayment"
                    name="minimumpayment"
                    value=""
                  >
                </div>
                <div>
                  <label for="account-1-nextduedate">
                    Next Due Date
                  </label>
                  <input
                    id="account-1-nextduedate"
                    name="nextduedate"
                    value=""
                  >
                </div>
              </div>
            </div>
            <div class="import-form-account">
              <div class="account-left">
                <input
                  id="account-2-id"
                  name="id"
                  type="hidden"
                  value="4f8978af-dc34-4c45-bfda-5fb82925cb93-unsynced"
                >
                <input
                  id="account-2-balance"
                  name="balance"
                  type="hidden"
                >
                <input
                  id="account-2-name"
                  name="name"
                  type="hidden"
                  value="Mortgage"
                >
                <div>
                  Mortgage
                </div>
                <div class="balance">
                  $125,789.00
                </div>
              </div>
              <div class="account-right">
                <div>
                  <label
                    class="import-form-label"
                    for="account-2-interestrate"
                  >
                    Interest Rate
                  </label>
                  <input
                    id="account-2-interestrate"
                    name="interestrate"
                    value=""
                  >
                </div>
                <div>
                  <label for="account-2-interestrate">
                    Min. Monthly Payment
                  </label>
                  <input
                    id="account-2-minimumpayment"
                    name="minimumpayment"
                    value=""
                  >
                </div>
                <div>
                  <label for="account-2-nextduedate">
                    Next Due Date
                  </label>
                  <input
                    id="account-2-nextduedate"
                    name="nextduedate"
                    value=""
                  >
                </div>
              </div>
            </div>
            <button
              class="btn btn-outline-purple"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
